File: app/javascript/controllers/azure_credentials_controller.js
import { Controller } from "@hotwired/stimulus"

// Azure credentials management — add, test, list, delete
export default class extends Controller {
    static targets = [
        "overlay", "list",
        // Add form fields
        "addOverlay", "addName", "addTenantId", "addClientId", "addClientSecret",
        "addType", "addExpDate", "addError", "addSaveBtn",
        // Test results
        "testOverlay", "testBody",
        // Delete confirm
        "deleteOverlay", "deleteCredName", "deleteCredId"
    ]
    static values = {
        listUrl: String,
        testUrl: String,
        createUrl: String,
        customerUuid: String,
        projectUuid: String
    }

    connect() {
        this.credentials = []
        this.loadCredentials()
    }

    // ── Data loading ──

    async loadCredentials() {
        try {
            const csrfToken = this.csrfToken()
            const resp = await fetch(this.listUrlValue, {
                headers: { "Accept": "application/json", "X-CSRF-Token": csrfToken }
            })
            const data = await resp.json()
            this.byTenant = data.by_tenant || {}
            this.renderList()
        } catch (e) {
            console.error("Failed to load credentials:", e)
        }
    }

    renderList() {
        if (!this.hasListTarget) return
        const tenantIds = Object.keys(this.byTenant)
        if (tenantIds.length === 0) {
            this.listTarget.innerHTML = ""
            return
        }

        let html = ""
        for (const [tenantId, creds] of Object.entries(this.byTenant)) {
            html += `<div class="card" style="margin-bottom:16px;">
                <div style="padding:10px 16px;background:var(--bg-elevated);border-bottom:1px solid var(--border);font-size:12px;color:var(--text-muted);font-weight:600;">
                    Tenant: <span class="mono" style="font-weight:400;color:var(--text-secondary);">${this.escapeHtml(tenantId)}</span>
                </div>
                <div style="padding:12px 16px;display:flex;flex-direction:column;gap:10px;">`

            creds.forEach(c => {
                const typeBadge = c.credential_type === "admin"
                    ? '<span class="badge badge-dev">Admin</span>'
                    : '<span class="badge badge-prd">Read-only</span>'
                const expLabel = c.expiration_date
                    ? `<span style="font-size:12px;color:var(--text-muted);">Expires: ${c.expiration_date}</span>`
                    : '<span style="font-size:12px;color:var(--text-muted);">No expiration</span>'
                const name = c.name || c.credential_type || "Credential"

                html += `<div style="border:1px solid var(--border);border-radius:8px;padding:12px 16px;">
                    <div style="display:flex;justify-content:space-between;align-items:center;">
                        <div style="display:flex;align-items:center;gap:10px;">
                            <span style="font-weight:600;font-size:14px;">${this.escapeHtml(name)}</span>
                            ${typeBadge}
                        </div>
                        <div style="display:flex;gap:6px;">
                            <button class="btn btn-ghost btn-sm" style="color:var(--red);font-size:12px;"
                                data-action="click->azure-credentials#openDelete"
                                data-cred-id="${c.credential_id || c.uuid || ''}"
                                data-cred-name="${this.escapeHtml(name)}">Delete</button>
                        </div>
                    </div>
                    <div style="display:flex;gap:20px;margin-top:8px;padding-top:8px;border-top:1px solid var(--border-light);font-size:13px;color:var(--text-muted);">
                        <span>Client: <span class="mono">${this.escapeHtml(c.client_id_masked || (c.client_id || "").substring(0, 8) + "…")}</span></span>
                        ${expLabel}
                        ${this.daysRemainingBadge(c.days_remaining)}
                    </div>
                </div>`
            })

            html += `</div></div>`
        }

        this.listTarget.innerHTML = html
    }

    // ── Add Modal ──

    openAdd() {
        this.addOverlayTarget.classList.add("open")
        document.body.style.overflow = "hidden"
        this.clearAddForm()
    }

    closeAdd() {
        this.addOverlayTarget.classList.remove("open")
        document.body.style.overflow = ""
    }

    clearAddForm() {
        if (this.hasAddNameTarget) this.addNameTarget.value = ""
        if (this.hasAddTenantIdTarget) this.addTenantIdTarget.value = ""
        if (this.hasAddClientIdTarget) this.addClientIdTarget.value = ""
        if (this.hasAddClientSecretTarget) this.addClientSecretTarget.value = ""
        if (this.hasAddExpDateTarget) this.addExpDateTarget.value = ""
        if (this.hasAddErrorTarget) this.addErrorTarget.textContent = ""
        if (this.hasAddSaveBtnTarget) this.addSaveBtnTarget.disabled = false
    }

    async testCredentials() {
        const tenantId = this.addTenantIdTarget.value.trim()
        const clientId = this.addClientIdTarget.value.trim()
        const clientSecret = this.addClientSecretTarget.value.trim()

        if (!tenantId || !clientId || !clientSecret) {
            this.addErrorTarget.textContent = "Tenant ID, Client ID, and Client Secret are required."
            return
        }

        this.addErrorTarget.textContent = ""

        try {
            const resp = await fetch(this.testUrlValue, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                    "Accept": "application/json",
                    "X-CSRF-Token": this.csrfToken()
                },
                body: JSON.stringify({ tenant_id: tenantId, client_id: clientId, client_secret: clientSecret })
            })
            const data = await resp.json()

            if (resp.ok && data.success) {
                this.showTestResults(true, data.data)
            } else {
                this.showTestResults(false, data.error || "Authentication failed")
            }
        } catch (e) {
            this.showTestResults(false, "Network error")
        }
    }

    showTestResults(success, detail) {
        if (!this.hasTestOverlayTarget) return
        this.testOverlayTarget.classList.add("open")

        if (success) {
            const subs = detail?.subscriptions || detail?.accessible_subscriptions || []
            let subsHtml = ""
            if (Array.isArray(subs) && subs.length > 0) {
                subsHtml = subs.map(s => {
                    const name = s.display_name || s.name || s.subscription_id
                    const id = s.subscription_id || ""
                    return `<div style="padding:8px 12px;border-bottom:1px solid var(--border-light);font-size:13px;">
                        <span style="color:var(--green);margin-right:6px;">✓</span>
                        <span style="font-weight:500;">${this.escapeHtml(name)}</span>
                        <span class="mono" style="color:var(--text-muted);margin-left:6px;font-size:11px;">(${this.escapeHtml(id)})</span>
                    </div>`
                }).join("")
            }
            this.testBodyTarget.innerHTML = `
                <div style="display:flex;align-items:center;gap:8px;margin-bottom:16px;">
                    <span style="color:var(--green);font-size:18px;">✓</span>
                    <span style="font-weight:600;color:var(--green);">Authentication Successful</span>
                </div>
                ${subsHtml ? `<div style="font-size:13px;font-weight:600;margin-bottom:8px;">Accessible Subscriptions:</div>
                <div style="border:1px solid var(--border);border-radius:8px;overflow:hidden;">${subsHtml}</div>` : ""}`
        } else {
            this.testBodyTarget.innerHTML = `
                <div style="display:flex;align-items:center;gap:8px;margin-bottom:16px;">
                    <span style="color:var(--red);font-size:18px;">✗</span>
                    <span style="font-weight:600;color:var(--red);">Authentication Failed</span>
                </div>
                <p style="font-size:13px;color:var(--text-muted);">${this.escapeHtml(String(detail))}</p>`
        }
    }

    closeTest() {
        this.testOverlayTarget.classList.remove("open")
    }

    async saveCredential() {
        const tenantId = this.addTenantIdTarget.value.trim()
        const clientId = this.addClientIdTarget.value.trim()
        const clientSecret = this.addClientSecretTarget.value.trim()

        if (!tenantId || !clientId || !clientSecret) {
            this.addErrorTarget.textContent = "Tenant ID, Client ID, and Client Secret are required."
            return
        }

        this.addSaveBtnTarget.disabled = true
        this.addSaveBtnTarget.textContent = "Saving…"
        this.addErrorTarget.textContent = ""

        const typeSelect = this.addOverlayTarget.querySelector('input[name="credential_type"]:checked')
        const body = {
            tenant_id: tenantId,
            client_id: clientId,
            client_secret: clientSecret,
            credential_type: typeSelect ? typeSelect.value : "read_only",
            name: this.addNameTarget.value.trim() || undefined,
            expiration_date: this.addExpDateTarget.value || undefined
        }

        try {
            const resp = await fetch(this.createUrlValue, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                    "Accept": "application/json",
                    "X-CSRF-Token": this.csrfToken()
                },
                body: JSON.stringify(body)
            })
            const data = await resp.json()

            if (resp.ok && data.success) {
                this.closeAdd()
                this.loadCredentials()
            } else {
                this.addErrorTarget.textContent = data.error || "Failed to save credential."
                this.addSaveBtnTarget.disabled = false
                this.addSaveBtnTarget.textContent = "Save"
            }
        } catch (e) {
            this.addErrorTarget.textContent = "Network error."
            this.addSaveBtnTarget.disabled = false
            this.addSaveBtnTarget.textContent = "Save"
        }
    }

    // ── Delete Modal ──

    openDelete(event) {
        const btn = event.currentTarget
        this.pendingDeleteId = btn.dataset.credId
        if (this.hasDeleteCredNameTarget) this.deleteCredNameTarget.textContent = btn.dataset.credName || "this credential"
        if (this.hasDeleteCredIdTarget) this.deleteCredIdTarget.value = btn.dataset.credId
        this.deleteOverlayTarget.classList.add("open")
    }

    closeDelete() {
        this.deleteOverlayTarget.classList.remove("open")
    }

    async confirmDelete() {
        if (!this.pendingDeleteId) return

        const url = this.createUrlValue.replace(/\/$/, "") + "/" + this.pendingDeleteId
        try {
            const resp = await fetch(url, {
                method: "DELETE",
                headers: {
                    "Accept": "application/json",
                    "X-CSRF-Token": this.csrfToken()
                }
            })
            if (resp.ok) {
                this.closeDelete()
                this.loadCredentials()
            }
        } catch (e) {
            console.error("Delete failed:", e)
        }
    }

    // ── Helpers ──

    backdropClose(event) {
        if (event.target === event.currentTarget) {
            event.target.classList.remove("open")
            document.body.style.overflow = ""
        }
    }

    closeOnEscape(event) {
        if (event.key === "Escape") {
            this.element.querySelectorAll(".modal-overlay.open").forEach(el => el.classList.remove("open"))
            document.body.style.overflow = ""
        }
    }

    csrfToken() {
        return document.querySelector('meta[name="csrf-token"]')?.content || ""
    }

    escapeHtml(str) {
        const div = document.createElement("div")
        div.textContent = str
        return div.innerHTML
    }

    daysRemainingBadge(days) {
        if (days === null || days === undefined) return ""
        if (days < 0) {
            return `<span class="badge" style="background:var(--red-dim);color:var(--red);">Expired ${Math.abs(days)}d ago</span>`
        }
        if (days <= 30) {
            return `<span class="badge" style="background:var(--orange-dim);color:var(--orange);">${days}d remaining</span>`
        }
        return `<span class="badge" style="background:var(--green-dim);color:var(--green);">${days}d remaining</span>`
    }
}

File: app/controllers/azure_credentials_controller.rb
# frozen_string_literal: true

class AzureCredentialsController < AuthenticatedController
  before_action :set_clients
  before_action :set_customer
  before_action :set_project

  # GET /customers/:customer_uuid/projects/:project_uuid/azure_credentials
  # JSON — returns credentials list for this customer
  def index
    response = @platform.list_azure(customer_uuid: @customer.uuid)
    if response.success?
      by_tenant = response.data.is_a?(Hash) ? (response.data[:by_tenant] || {}) : {}
      render json: { by_tenant: by_tenant }
    else
      render json: { by_tenant: {}, error: response.error }, status: :unprocessable_entity
    end
  end

  # POST /customers/:customer_uuid/projects/:project_uuid/azure_credentials/test
  # JSON — tests Azure credentials (raw or stored by credential_id)
  def test
    if params[:credential_id].present?
      # Test a stored credential — look up secret from SSM and test
      Rails.logger.info "Testing stored credential: customer=#{@customer.uuid}, tenant=#{params[:tenant_id]}, cred=#{params[:credential_id]}"
      result = @platform.get_azure_with_secret(
        customer_uuid: @customer.uuid,
        tenant_id: params[:tenant_id].to_s.strip,
        credential_id: params[:credential_id].to_s.strip
      )
    else
      result = @platform.test_azure(
        tenant_id:   params[:tenant_id].to_s.strip,
        client_id:   params[:client_id].to_s.strip,
        client_secret: params[:client_secret].to_s.strip
      )
    end

    if result.success?
      data = result.data.is_a?(Hash) ? result.data : {}
      render json: { success: true, data: data }
    else
      error_msg = result.error || result.data
      if error_msg.is_a?(Hash)
        error_msg = error_msg[:detail] || error_msg["detail"] || error_msg.to_s
      end
      render json: { success: false, error: error_msg || "Authentication failed" }, status: :unprocessable_entity
    end
  end

  # POST /customers/:customer_uuid/projects/:project_uuid/azure_credentials
  # JSON — saves a credential
  def create
    authorize!(:manage, @project)

    attrs = {
      customer_uuid:   @customer.uuid,
      tenant_id:       params[:tenant_id].to_s.strip,
      client_id:       params[:client_id].to_s.strip,
      client_secret:   params[:client_secret].to_s.strip,
      credential_type: params[:credential_type] || "read_only",
      name:            params[:name].to_s.strip.presence,
      expiration_date: params[:expiration_date].to_s.strip.presence
    }.compact

    response = @platform.save_azure(**attrs)
    if response.success?
      render json: { success: true, data: response.data }
    else
      render json: { success: false, error: response.error || response.data }, status: :unprocessable_entity
    end
  end

  # PATCH /customers/:customer_uuid/projects/:project_uuid/azure_credentials/:id
  # JSON — updates a credential
  def update
    authorize!(:manage, @project)

    attrs = {
      customer_uuid: @customer.uuid,
      credential_id: params[:id]
    }
    attrs[:name]            = params[:name].to_s.strip            if params[:name].present?
    attrs[:expiration_date] = params[:expiration_date].to_s.strip if params[:expiration_date].present?
    attrs[:client_id]       = params[:client_id].to_s.strip       if params[:client_id].present?
    attrs[:client_secret]   = params[:client_secret].to_s.strip   if params[:client_secret].present?

    response = @platform.update_azure(**attrs)
    if response.success?
      render json: { success: true, data: response.data }
    else
      render json: { success: false, error: response.error || response.data }, status: :unprocessable_entity
    end
  end

  # DELETE /customers/:customer_uuid/projects/:project_uuid/azure_credentials/:id
  # JSON — deletes a credential
  def destroy
    authorize!(:manage, @project)

    response = @platform.delete_azure(
      customer_uuid: @customer.uuid,
      credential_id: params[:id]
    )
    if response.success?
      render json: { success: true }
    else
      render json: { success: false, error: response.error || response.data }, status: :unprocessable_entity
    end
  end

  # POST /customers/:customer_uuid/projects/:project_uuid/azure_credentials/expiring
  # JSON — returns expiring credentials
  def expiring
    response = @platform.expiring_azure(customer_uuid: @customer.uuid)
    if response.success?
      render json: { credentials: Array(response.data) }
    else
      render json: { credentials: [], error: response.error }, status: :unprocessable_entity
    end
  end

  private

  def set_clients
    @client   = CsInternalApiClient.new
    @platform = PlatformApiClient.new
  end

  def set_customer
    response = @client.get_customer(uuid: params[:customer_uuid])
    if response.success?
      @customer = Customer.from_api(response.data)
    else
      redirect_to customers_path, alert: "Customer not found."
    end
  end

  def set_project
    response = @client.get_project(uuid: params[:project_uuid])
    if response.success?
      @project = Project.from_api(response.data)
    else
      redirect_to customer_path(params[:customer_uuid]), alert: "Project not found.", only_path: true
    end
  end
end
